utils: add reproducible_gzip helper function

alebcay opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Introduce a new `reproducible_gzip` helper function which can be used by formulae. The inputs and outputs of this function are the same as `Utils.gzip`. However, the resulting gz file produced by `reproducible_gzip` has its `mtime` set to `SOURCE_DATE_EPOCH`.

This allows for gz files produced at different times, on different CI instances or end user machines, and with different operating systems to still have the same checksum.

So far, I have identified five formulae in `homebrew-core` that the bottles probably only differ due to different mtimes in gzipped files (either manpages or data files). Introducing this usage in those instances should allow an `all` bottle to be built.

Notes and other considerations:
- As noted in the comment in `Utils.reproducible_gzip`, this currently requires `SOURCE_DATE_EPOCH` env var to be set (which it should be in the formula build environment). I originally planned to use an mtime of 0 if `SOURCE_DATE_EPOCH` is not set, but there's a bug in the Zlib gem which has been fixed but doesn't seem to have made it to us yet.
- The Gzip buffer size of `64 * 1024` is currently left as a magic number in the function. I'm open to making it a constant at the top of the file if that's preferred. Note that the same constant is also used in https://github.com/Homebrew/brew/blob/master/Library/Homebrew/dev-cmd/bottle.rb, so not sure if it would make sense to unify the constant somewhere else.
- I've implemented this as a separate function from `Utils.gzip` to avoid changing the existing functionality, and since this imposes additional requirements for usage (i.e. `SOURCE_DATE_EPOCH` must be set). I'm also open to simply amending the existing `Utils.gzip` function with this feature if that's preferred (i.e. non-reproducible behavior if `SOURCE_DATE_EPOCH` is not set; reproducible behavior if it is set).

MikeMcQuaid commented on Library/Homebrew/utils.rb:
```diff
@@ -404,6 +404,30 @@ def gzip(*paths)
+    #
+    # This was fixed in https://github.com/ruby/zlib/pull/10. Set mtime to 0 instead
+    # of raising exception once we are using zlib gem version 1.1.0 or newer.
+    no_env_var_message = "Can't create reproducible gzip file without SOURCE_DATE_EPOCH environment variable set"
+    raise no_env_var_message if ENV["SOURCE_DATE_EPOCH"].blank?
```
```suggestion
    if ENV["SOURCE_DATE_EPOCH"].blank?
      raise ArgumentError, 
        "Can't create reproducible gzip file without SOURCE_DATE_EPOCH environment variable set"
    end
```

MikeMcQuaid commented on Library/Homebrew/utils.rb:
```diff
@@ -404,6 +404,30 @@ def gzip(*paths)
     end
   end
 
+  def reproducible_gzip(*paths)
```
Feels a little odd to take multiple `paths` here rather than a single one. Thoughts?

Also, might be nice to add a Sorbet signature here.

alebcay replied:
I'd be fine with this just accepting a single path rather than arbitrarily many. Some formulae do pass in an array of paths already so to transition them to this new function would involve adding an `.each` loop in the formulae. Not a big deal but just wanted to mention that the multiple `paths` form of the existing `gzip` function _is_ currently being used.

MikeMcQuaid commented on Library/Homebrew/utils.rb:
```diff
@@ -404,6 +404,30 @@ def gzip(*paths)
+      gz.orig_name = File.basename(path)
+
+      File.open(path, "rb") do |fp|
+        gz.write(fp.read(64 * 1024)) until fp.eof?
```
Would be good to add a comment explaining why these magic numbers are being used and pull them into a variable outside the loop.

MikeMcQuaid commented on Library/Homebrew/utils.rb:
```diff
@@ -404,6 +404,30 @@ def gzip(*paths)
+    raise no_env_var_message if ENV["SOURCE_DATE_EPOCH"].blank?
+
+    paths.map do |path|
+      gz = Zlib::GzipWriter.open("#{path}.gz")
```
Could make a variable for the `"#{path}.gz"`

MikeMcQuaid commented on Library/Homebrew/utils.rb:
```diff
@@ -404,6 +404,30 @@ def gzip(*paths)
+
+    paths.map do |path|
+      gz = Zlib::GzipWriter.open("#{path}.gz")
+      gz.mtime = ENV["SOURCE_DATE_EPOCH"].to_i
```
Similarly, could extract `ENV["SOURCE_DATE_EPOCH"].to_i` to a variable outside the loop.

MikeMcQuaid reviewed (commented):
Looks good so far! Would like to see `bottle.rb` modified to use this logic before this is merged to avoid two mostly duplicate implementations.

alebcay commented:
Made some changes based on the feedback.
- Changed the arguments to accept only a single path. The preferred `mtime` and `orig_name` attributes of the gz file are also exposed as optional keyword parameters. The `output` file location is also exposed as an optional keyword parameter.
- Changed the bottling process in `bottle.rb` to use `Utils.reproducible_hash`. It takes advantage of the newly exposed optional keyword parameters mentioned above.
- Moved the gzip buffer size `64 * 1024` into a variable in `Utils.reproducible_hash`

Outstanding concerns:
- Not sure how to explain the choice of `64 * 1024` as the gzip buffer size. Frankly, I stuck with it because it's what was already there. Not sure if changing it may affect the resulting gz file's checksum.
- I tried adding a `sig` block before the new `Utils.reproducible_hash` function and enabling `typed: true` at the top of the file, but that seems to create a lot of other errors with `brew typecheck`. Is there some way to enable typing just for that function? Or is it sufficient to just add the `sig` block there and other places that call it (where `typed: true`) will enforce the typechecking?

MikeMcQuaid commented:
> * Not sure how to explain the choice of `64 * 1024` as the gzip buffer size. Frankly, I stuck with it because it's what was already there. Not sure if changing it may affect the resulting gz file's checksum.

I think @Bo98 came up with this and may be able to elaborate?

> * Or is it sufficient to just add the `sig` block there and other places that call it (where `typed: true`) will enforce the typechecking?

I think it's sufficient to just add it.

MikeMcQuaid commented on Library/Homebrew/dev-cmd/bottle.rb:
```diff
@@ -425,14 +424,8 @@ def bottle_formula(f, args:)
-          end
-          gz.close
-          rm_f relocatable_tar_path
+          reproducible_gzip(relocatable_tar_path, mtime: tab.source_modified_time, orig_name: relocatable_tar_path,
+                            output: bottle_path)
```
```suggestion
          reproducible_gzip(relocatable_tar_path, 
                            mtime: tab.source_modified_time, 
                            orig_name: relocatable_tar_path,
                            output: bottle_path)
```

MikeMcQuaid commented on Library/Homebrew/utils.rb:
```diff
@@ -404,6 +404,36 @@ def gzip(*paths)
     end
   end
 
+  def reproducible_gzip(path, mtime: ENV["SOURCE_DATE_EPOCH"].to_i, orig_name: File.basename(path),
```
May even want to create a new `Utils::Gzip` or `Utils::Reproducible` just for this and call `Utils::Gzip.reproducible_gzip` as a more explicit API. In an ideal world we'd probably not rely on this `utils` global include.

If desired could also have a `reproducible_gzips` helper or something that does the looping, if it'll make the homebrew/core formulae usage nicer.

MikeMcQuaid reviewed (commented):
Great work here! Sorry for all the back and forth, loving how this is looking.

Bo98 commented:
> > * Not sure how to explain the choice of `64 * 1024` as the gzip buffer size. Frankly, I stuck with it because it's what was already there. Not sure if changing it may affect the resulting gz file's checksum.
> 
> I think @Bo98 came up with this and may be able to elaborate?

Apple's gzip also uses zlib so I made the buffer size the same for consistency reasons.

https://github.com/apple-oss-distributions/file_cmds/blob/file_cmds-400/gzip/gzip.c#L147

I didn't actually test if changing the buffer size affected checksums or not but there was no real reason to pick any other value.

Ultimately the primary goal was a consistent checksum regardless of the OS which Ruby's zlib allows us to do. The ability to match the checksum with what we previously had under Apple's gzip was extra motivation and made it an obvious choice.

MikeMcQuaid commented:
> Apple's gzip also uses zlib so I made the buffer size the same for consistency reasons.
> 
> https://github.com/apple-oss-distributions/file_cmds/blob/file_cmds-400/gzip/gzip.c#L147

Good to know, this would be a great Ruby comment here.

MikeMcQuaid commented on Library/Homebrew/utils.rb:
```diff
@@ -404,6 +404,36 @@ def gzip(*paths)
+            "Can't create reproducible gzip file without a valid mtime"
+    end
+
+    gzip_buffer_size = 64 * 1024
```
```suggestion
    # Apple's gzip also uses zlib so use the same buffer size here.
    # https://github.com/apple-oss-distributions/file_cmds/blob/file_cmds-400/gzip/gzip.c#L147
    gzip_buffer_size = 64 * 1024
```

alebcay commented:
Implemented the suggestions above and moved all of the new logic into `Utils::Gzip` module. I also moved the body of the existing `Utils.gzip` function to `Utils::Gzip.gzip`, leaving `Utils.gzip` present for compatibility, since it seemed to make sense to have all the `gzip`-related helper functions in one place. Let me know if it would be preferable not to move the existing `Utils.gzip` function.

One question about `require` statements in new `Utils` modules - I found that what I have above seems to pass tests and work in `brew irb`, although I am not explicitly requiring anything to get `zlib`, `safe_system`, etc. Not sure if I should explicitly `require` what's needed (and in that case, how to even identify what I need to `require`) or if it's fine as-is.

MikeMcQuaid commented:
> if it's fine as-is.

Fine as-is if it works!

MikeMcQuaid reviewed (approved):
Looks good! One tiny nit. Just to check the API use here: can I see a homebrew/core PR that will use this new API? Obviously it'll fail until this PR is merged but would be good to view the usage.

alebcay commented:
Sure, https://github.com/Homebrew/homebrew-core/pull/119509 is an example usage.

MikeMcQuaid commented:
Thanks @alebcay! I guess that makes me wonder: why have the `reproducible_gzip` and `gzip` as separate methods at all? What situations would you want to have a non-reproducible gzip? I wonder if in that situation you could just do `gzip(reproducible: false)` or `gzip(system_gzip: true)` or something?

alebcay commented:
> Thanks @alebcay! I guess that makes me wonder: why have the `reproducible_gzip` and `gzip` as separate methods at all? What situations would you want to have a non-reproducible gzip? I wonder if in that situation you could just do `gzip(reproducible: false)` or `gzip(system_gzip: true)` or something?

That's fair, I did consider that possibility too. I mostly left the existing `gzip` function alone for compatibility purposes only. Unifying them is probably possible, but the parameter differences between `reproducible_gzip` and `gzip` make the parameter handling a bit weird.

I think it would be easier to unify `reproducible_gzips` and `gzip` since they both handle multiple files. `reproducible_gzip` allows an optional `output` and `orig_name` parameter to be set when processing a single file.

We could allow these parameters when handling multiple files with `reproducible_gzips`, but so far there's no use case and it's complex enough that there's little benefit over the caller just using `.each` over single-file `reproducible_gzip` if that level of control is needed.

So I think I'll try combining the multi-file forms `reproducible_gzips` and `gzip`. I think it would be best to rename the `reproducible_gzip` function in this case to something that indicates that it specifically handles one file but allows more specificity with options (`orig_name` and `output` parameters) - open to name suggestions on this.

MikeMcQuaid commented:
> I think it would be easier to unify `reproducible_gzips` and `gzip` since they both handle multiple files.

Sounds good 👍🏻

> pen to name suggestions on this.

`gzip_with_options`?

alebcay commented:
Updated the PR with the suggested changes.
- `Utils::Gzip.gzip_with_options` now holds the logic for producing a reproducible gz file.
- `Utils::Gzip.gzip` creates one or multiple gz files, reproducible by default. A custom `mtime` can be passed if the env variable `SOURCE_DATE_EPOCH` is unavailable or shouldn't be used.
  - There's no logic checking if `reproducible` is false and `mtime` is set (in this case the `mtime` is ignored since we're not making a reproducible gz).
- `Utils.gzip` remains for compatibility purposes with existing formulae. The function signature is unchanged but it will build reproducible gz files. We could add the same optional params that were introduced in `Utils::Gzip.gzip` and pass those through, but so far there's no use case in formulae to set a custom `mtime` or _not_ build a reproducible gz.

Possible and/or future areas for improvement:
- Maybe `Utils.gzip` can be deprecated and removed in the future after formulae move to `Utils::Gzip.gzip` for some time? Assuming that it's desirable to move stuff out of top-level `Utils`.
- Does `Utils::Gzip.compress` and `Utils::Gzip.compress_with_options` sound more readable/human-friendly (and less redundant)? Open to this change if preferred.

MikeMcQuaid commented:
> * Does `Utils::Gzip.compress` and `Utils::Gzip.compress_with_options` sound more readable/human-friendly (and less redundant)? Open to this change if preferred, or wondering if there's some way to make it look less redundant when used in formulae.

Yeh, I like this naming!



> * Maybe `Utils.gzip` can be deprecated and removed in the future after formulae move to `Utils::Gzip.gzip` for some time? Assuming that it's desirable to move stuff out of top-level `Utils`.

Agreed. Can you add a `# odeprecated` comment in there and it'll be picked up and done on the next major or minor release?

alebcay commented:
This is hopefully all set now.

MikeMcQuaid reviewed (approved):
Looks good to me! Sorry for all the back and forth and thanks for your patience here @alebcay. Note we'll need to wait until this lands in a tag (most likely a week today) before we merge homebrew-core PRs relying on the new APIs but we (and should!) start testing this on homebrew-core PRs nowish. Feel free to CC me there!

alebcay commented:
Thanks for your guidance on organizing this change. Appreciate your patience as I weigh a lot of different ways to implement this.

I've opened https://github.com/Homebrew/homebrew-core/pull/120429 which is a first batch of changes. These should directly benefit from having identical bottles across all runners.

Other usages of `gzip` should be transitioned over but still will not produce identical bottles even with this change (i.e. bottles contain some other compiled/platform-specific binaries/files).

MikeMcQuaid commented:
Thanks @alebcay, looks great. Have enabled auto-merge and rerun flaky test. Nag me if something needs kicked again!

Bo98 commented on Library/Homebrew/utils/gzip.rb:
```diff
@@ -0,0 +1,73 @@
+module Utils
+  # Helper functions for creating gzip files.
+  #
+  # @api private
+  module Gzip
```
We don't seem to be treating this as private API? We deprecated the plain `gzip` Kernel extension function, pointed people to the (fully-qualified) `Utils::Gzip.compress` and have made changes to homebrew-core to use it.

MikeMcQuaid replied:
Yes, sounds like this should be public.
